src/rgw/rgw_dedup_store.cc

```diff
@@ -0,0 +1,847 @@
+                                   record_info_t          *p_rec_info) // OUT-PARAM
+  {
+    if (unlikely(!p_arr)) {
+      report_bad_access(dpp, __func__);
+      return -1;
+    }
+
+    ldpp_dout(dpp, 20) << __func__  << "::worker_id=" << (uint32_t)d_worker_id
+                       << ", md5_shard=" << (uint32_t)d_md5_shard
+                       << "::" << p_bucket->get_name() << "/" << obj_name << dendl;
+    disk_record_t rec;
+    int ret = fill_disk_record(&rec, p_bucket, p_obj, p_parsed_etag, obj_name, obj_size);
+    if (unlikely(ret != 0)) {
+      return ret;
+    }
+    p_rec_info->has_shared_manifest = rec.has_shared_manifest();
+    p_rec_info->has_valid_sha256    = rec.has_valid_sha256();
+    p_stats->egress_records ++;
+    // first, try and add the record to the current open block
+    p_rec_info->rec_id = p_curr_block->add_record(&rec, dpp);
+    if (p_rec_info->rec_id < MAX_REC_IN_BLOCK) {
+      p_rec_info->block_id = p_curr_block->get_block_id();
+      return 0;
+    }
+    else {
+      // Not enough space left in current block, close it and open the next block
+      ldpp_dout(dpp, 20) << __func__ << "::Block is full-> close and move to next" << dendl;
+      p_stats->egress_blocks++;
+      p_curr_block->close_block(dpp, true);
+    }
+
+    // Do we have more Blocks in the block-array ?
+    if (p_curr_block < last_block()) {
+      p_curr_block ++;
+      d_seq_number ++;
+      p_curr_block->init(d_worker_id, d_seq_number);
+      p_rec_info->rec_id = p_curr_block->add_record(&rec, dpp);
+    }
+    else {
+      ldpp_dout(dpp, 20)  << __func__ << "::calling flush()" << dendl;
+      ret = flush(ioctx);
+      p_rec_info->rec_id = p_curr_block->add_record(&rec, dpp);
+    }
+
+    p_rec_info->block_id = p_curr_block->get_block_id();
+    return ret;
+  }
+
+  //---------------------------------------------------------------------------
+  disk_block_array_t::disk_block_array_t(const DoutPrefixProvider* dpp,
+                                         uint8_t *raw_mem,
+                                         uint64_t raw_mem_size,
+                                         work_shard_t worker_id,
+                                         worker_stats_t *p_worker_stats,
+                                         work_shard_t num_work_shards,
+                                         md5_shard_t num_md5_shards)
+  {
+    d_num_md5_shards = num_md5_shards;
+    d_worker_id = worker_id;
+    disk_block_t *p     = (disk_block_t *)raw_mem;
+    disk_block_t *p_end = (disk_block_t *)(raw_mem + raw_mem_size);
+
+    for (unsigned md5_shard = 0; md5_shard < d_num_md5_shards; md5_shard++) {
+      ldpp_dout(dpp, 20) << __func__ << "::p=" << p << "::p_end=" << p_end << dendl;
+      if (p + DISK_BLOCK_COUNT <= p_end) {
+        d_disk_arr[md5_shard].activate(dpp, p, d_worker_id, md5_shard,
+                                       p_worker_stats);
+        p += DISK_BLOCK_COUNT;
+      }
+      else {
```

Comment:
why would an overflow happen?
if this is a software bug, we need to crash or assert on it.
if this can be from user input, invalid value on disk, etc. we should detect it in a much earlier stage and make sure the error is handled there